File: track-b-data-engineering/parser/fitment_resolver.py
"""Fitment resolver — Python port of fitment-resolver.ts.

Year/make/model fitment is encoded in the SHEET NAME (the data file
doesn't have explicit fitment columns). Returns ResolvedFitment when at
least year OR model code can be extracted; returns None for sheets like
"TABLE OF CONTENTS", "CARBURETOR JETS", etc.
"""

from __future__ import annotations

import re
from dataclasses import dataclass


@dataclass(frozen=True)
class ResolvedFitment:
    year_start: int | None
    year_end: int | None
    model_code: str | None
    variant: str | None


_MODEL_CODE_RE = re.compile(
    r"\b(AY|AT|AU|KMB|TS|TSD|TD|TT|K2|K4|K6|KT|T2|T4|S70|S200|S350|eA|eKMB)\d*-?\d*[A-Z]?\b",
    re.IGNORECASE,
)
_RANGE_PAREN_RE = re.compile(r"\((\d{4})\s*-\s*(\d{4})\)")
_OPEN_END_PAREN_RE = re.compile(r"\((\d{4})\s*\+\)")
_SINGLE_YEAR_PAREN_RE = re.compile(r"\((\d{4})\)")
_OPEN_END_PREFIX_RE = re.compile(r"\b(\d{4})\+")
_VARIANT_RE = re.compile(r"\b(EPA|EFI)\b", re.IGNORECASE)


def resolve_fitment_from_sheet_name(sheet_name: str) -> ResolvedFitment | None:
    name = sheet_name.strip()
    if _is_non_fitment_sheet(name):
        return None

    year_start: int | None = None
    year_end: int | None = None

    range_m = _RANGE_PAREN_RE.search(name)
    if range_m:
        year_start = int(range_m.group(1))
        year_end = int(range_m.group(2))
    else:
        open_end = _OPEN_END_PAREN_RE.search(name)
        if open_end:
            year_start = int(open_end.group(1))
            year_end = None
        else:
            single = _SINGLE_YEAR_PAREN_RE.search(name)
            if single:
                year_start = int(single.group(1))
                year_end = int(single.group(1))
            else:
                prefix = _OPEN_END_PREFIX_RE.search(name)
                if prefix:
                    year_start = int(prefix.group(1))
                    year_end = None

    model_match = _MODEL_CODE_RE.search(name)
    model_code = model_match.group(0).upper() if model_match else None

    variant_match = _VARIANT_RE.search(name)
    variant = variant_match.group(1).upper() if variant_match else None

    if year_start is None and model_code is None:
        return None

    return ResolvedFitment(
        year_start=year_start,
        year_end=year_end,
        model_code=model_code,
        variant=variant,
    )


def _is_non_fitment_sheet(name: str) -> bool:
    upper = name.upper()
    if upper in {
        "TABLE OF CONTENTS",
        "TOC",
        "SHEET18",
        "CARBURETOR JETS",
        "FORK SEAL SPECS",
        "ATV WHEEL SPECS",
        "SPOKE SPECS",
        "SPARK PLUGS",
        "BATTERY SPECS",
        "OWNERS MANUALS",
        "SNOW TRACK KIT",
        "SKI KIT PARTS",
        "EC2I",
    }:
        return True
    if upper.startswith("DIRTBIKEPITBIKE WHEEL BEARING"):
        return True
    if "UPGRADE KIT" in upper:
        return True
    return False

File: track-a-jd-native/src/ingest/fitment-resolver.ts
/**
 * Fitment resolver.
 *
 * Year/make/model fitment is encoded in the SHEET NAME (the data file
 * doesn't have explicit fitment columns). Examples observed in the sample:
 *
 *   "FOXStorm 70 AY70-2 "              → year_start=null, year_end=null,
 *                                         model_code="AY70-2"
 *   "PREDATOR 125 (2016-2020)"         → 2016..2020, model_code="AT125-B"
 *   "Bull 180 AU180 (2020-2022)"       → 2020..2022, model_code="AU180"
 *   "AU180-2 (2023+) Parts Diagram"    → 2023..null, model_code="AU180-2"
 *   "2024+ TT125 EPA"                  → 2024..null, model_code="TT125", variant="EPA"
 *
 * Heuristics:
 *   • Year ranges in parens: "(YYYY-YYYY)" | "(YYYY+)" | "(YYYY)"
 *   • Year prefix: "YYYY+ ..." | "YYYY ..."
 *   • Model code regex: well-known prefixes from the OEM
 *   • Variant: "EPA" | "EFI" | "D" appearing as a separate word
 *
 * Returns null when no fitment information can be extracted (eg "TABLE OF
 * CONTENTS", "Sheet18", "SPARK PLUGS"). The caller decides whether to
 * skip the sheet or treat it as a reference/spec sheet.
 */

export interface ResolvedFitment {
  yearStart: number | null;
  yearEnd: number | null;
  modelCode: string | null;
  variant: string | null;
  // We don't infer `make` here — that's a per-dealer config from `rules.yaml`.
  // The caller adds `make: "Kayo"` (or whatever) before persisting.
}

const MODEL_CODE_RE = /\b(AY|AT|AU|KMB|TS|TSD|TD|TT|K2|K4|K6|KT|T2|T4|S70|S200|S350|eA|eKMB)\d*-?\d*[A-Z]?\b/i;
const RANGE_PAREN_RE = /\((\d{4})\s*-\s*(\d{4})\)/;
const OPEN_END_PAREN_RE = /\((\d{4})\s*\+\)/;
const SINGLE_YEAR_PAREN_RE = /\((\d{4})\)/;
const OPEN_END_PREFIX_RE = /\b(\d{4})\+/;
// Variant only matches standalone "EPA" / "EFI" markers.
// "D" is ambiguous (suffix on model codes like AU125-D), so we don't
// treat it as a separate variant — it stays as part of model_code.
const VARIANT_RE = /\b(EPA|EFI)\b/i;

export function resolveFitmentFromSheetName(
  sheetName: string,
): ResolvedFitment | null {
  const name = sheetName.trim();

  // Skip non-fitment sheets fast.
  if (isNonFitmentSheet(name)) return null;

  // Year window.
  let yearStart: number | null = null;
  let yearEnd: number | null = null;

  const rangeMatch = name.match(RANGE_PAREN_RE);
  if (rangeMatch) {
    yearStart = Number(rangeMatch[1]);
    yearEnd = Number(rangeMatch[2]);
  } else {
    const openEndParen = name.match(OPEN_END_PAREN_RE);
    if (openEndParen) {
      yearStart = Number(openEndParen[1]);
      yearEnd = null;
    } else {
      const singleParen = name.match(SINGLE_YEAR_PAREN_RE);
      if (singleParen) {
        yearStart = Number(singleParen[1]);
        yearEnd = Number(singleParen[1]);
      } else {
        const openEndPrefix = name.match(OPEN_END_PREFIX_RE);
        if (openEndPrefix) {
          yearStart = Number(openEndPrefix[1]);
          yearEnd = null;
        }
      }
    }
  }

  // Model code.
  const modelMatch = name.match(MODEL_CODE_RE);
  const modelCode = modelMatch?.[0]?.toUpperCase() ?? null;

  // Variant.
  const variantMatch = name.match(VARIANT_RE);
  const variant = variantMatch?.[1]?.toUpperCase() ?? null;

  // If we got neither a year nor a model code, we don't have enough to call it fitment.
  if (yearStart === null && modelCode === null) return null;

  return { yearStart, yearEnd, modelCode, variant };
}

function isNonFitmentSheet(name: string): boolean {
  const upper = name.toUpperCase();
  return (
    upper === "TABLE OF CONTENTS" ||
    upper === "TOC" ||
    upper === "SHEET18" ||
    upper === "CARBURETOR JETS" ||
    upper === "FORK SEAL SPECS" ||
    upper === "ATV WHEEL SPECS" ||
    upper === "SPOKE SPECS" ||
    upper === "SPARK PLUGS" ||
    upper === "BATTERY SPECS" ||
    upper === "OWNERS MANUALS" ||
    upper.startsWith("DIRTBIKEPITBIKE WHEEL BEARING") ||
    upper === "SNOW TRACK KIT" ||
    upper === "SKI KIT PARTS" ||
    upper.includes("UPGRADE KIT") ||
    upper === "EC2I"
  );
}
